File: 383. Ransom Note/main.cpp
#include <vector>
#include <iostream>
#include <unordered_map>

class Solution {
public:
    bool canConstruct(std::string ransomNote, std::string magazine) {
        // construct a map of char count in magazine
        std::unordered_map<char, int> chars;
        for(char c : magazine) chars[c]++;
        
        // go through ransomNote subtracting char counts
        for(char c : ransomNote){
            if(chars[c] > 0) chars[c]--;
            else return false;
        }

        return true;
    }
};

int main()
{
    std::string ran = "aba";
    std::string mag = "baa";
    Solution s = Solution();
    auto res = s.canConstruct(ran, mag);
    std::cout << res;
    return 0;
}

File: 383. Ransom Note/main.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        for a in ransomNote:
            if a in magazine:
                i = magazine.find(a)
                magazine = magazine[:i] + magazine[i+1:]
                
            else: return False

        return True



def main():
    s = Solution()
    ran = "aa"
    mag = "ab"
    print(s.canConstruct(ran, mag))


if __name__ == "__main__":
    main()